File: application/src/main/java/whenyourcar/application/facade/car/CarSaleFacade.java
package whenyourcar.application.facade.car;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import whenyourcar.application.service.car.CarSaleViewService;
import whenyourcar.domain.entity.CarSale;
import whenyourcar.domain.entity.User;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import whenyourcar.application.dto.car.sale.CarSaleRequest;
import whenyourcar.application.dto.car.search.CarCommonResponse;
import whenyourcar.application.service.car.CarSaleService;
import whenyourcar.application.service.user.UserCommonService;

import java.util.Date;

@Service
@RequiredArgsConstructor
public class CarSaleFacade {
    private final CarSaleService carSaleService;
    private final UserCommonService userCommonService;
    private final CarSaleViewService carSaleViewService;

    public void postSaleCar(CarSaleRequest.CarSaleRequestDto carSaleRequest,
                            String email) {
        User user = userCommonService.getUserByEmail(email);
        carSaleService.postSaleCar(carSaleRequest, user);
    }

    public Page<CarCommonResponse.SearchResponseDto> searchCars(Pageable pageable, Date minAge, Date maxAge, Integer minMileage, Integer maxMileage, Integer minPrice, Integer maxPrice, String color) {
        return carSaleService.searchCarsService(pageable, minAge, maxAge, minMileage, maxMileage, minPrice, maxPrice,color);
    }

    public CarCommonResponse.SearchDescriptionSaleResponseDto searchDescription(Long carId,
                                                                                HttpServletRequest httpServletRequest,
                                                                                HttpServletResponse httpServletResponse) {
        CarSale carSale = carSaleService.findCarSaleById(carId);
        httpServletResponse.addCookie(carSaleViewService.increaseViewCount(carSale.getCarId(), httpServletRequest));
        return carSaleService.searchDescriptionService(carSale);
    }

    public Page<CarCommonResponse.SearchDetailResponseDto> searchDetailCars(Pageable pageable, String manu, String model, String submodel, String grade ) {
        return carSaleService.searchDetailCarsService(pageable, manu, model, submodel, grade);
    }
}

File: application/src/main/kotlin/howmuchcar/application/facade/car/CarSaleFacade.kt
package howmuchcar.application.facade.car

import com.fasterxml.jackson.databind.ObjectMapper
import howmuchcar.application.converter.car.CarSaleSearchDescResponse
import howmuchcar.application.converter.car.CarSearchDetailResponse
import howmuchcar.application.converter.car.CarSearchResponse
import howmuchcar.application.dto.car.CarSaleRequest
import howmuchcar.application.port.`in`.car.CarSaleCommonUseCase
import howmuchcar.application.port.`in`.car.CarSaleSearchUseCase
import howmuchcar.application.port.`in`.car.CarSaleViewUseCase
import howmuchcar.application.port.out.infra.BucketPort
import howmuchcar.application.port.out.infra.EmbeddingPort
import howmuchcar.application.port.out.infra.GenerateTagPort
import howmuchcar.application.port.out.infra.PineconePort
import howmuchcar.domain.entity.CarSale
import howmuchcar.domain.entity.User
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import lombok.RequiredArgsConstructor
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.web.multipart.MultipartFile
import java.time.LocalDate
import java.util.*

@Service
@RequiredArgsConstructor
class CarSaleFacade (
    private val carSaleCommonUseCase: CarSaleCommonUseCase,
    private val carSaleSearchUseCase: CarSaleSearchUseCase,
    private val carSaleViewUseCase: CarSaleViewUseCase,
    private val bucketPort: BucketPort,
    private val embeddingPort: EmbeddingPort,
    private val generateTagPort: GenerateTagPort,
    private val pineconePort: PineconePort,
    private val objectMapper: ObjectMapper
){
    fun postSaleCar(
        carSaleRequest: CarSaleRequest,
        user: User,
        images: List<MultipartFile>
    ) {
        val imagesURLs: MutableList<String> = ArrayList()
        for (image in images) {
            imagesURLs.add(bucketPort.uploadFile(image))
        }
        val tags = generateTagPort.generateTags(carSaleRequest.description)
        val carSale = carSaleCommonUseCase.postSaleCar(carSaleRequest, user, imagesURLs, objectMapper.writeValueAsString(tags))

        pineconePort.upsertVectors(embeddingPort.generateEmbedding(tags), carSale.carId)
    }

    fun patchCarToSaleCompleted(carId: Long, user: User) {
        carSaleCommonUseCase.patchCarToSaleCompleted(carId, user)
    }

    fun searchCars(
        pageable: Pageable,
        minAge: LocalDate?,
        maxAge: LocalDate?,
        minMileage: Int?,
        maxMileage: Int?,
        minPrice: Int?,
        maxPrice: Int?,
        color: String?
    ): Page<CarSearchResponse> {
        return carSaleSearchUseCase.searchCarSale(
            pageable,
            minAge,
            maxAge,
            minMileage,
            maxMileage,
            minPrice,
            maxPrice,
            color
        )
    }

    fun searchDescription(
        carId: Long,
        httpServletRequest: HttpServletRequest,
        httpServletResponse: HttpServletResponse
    ): CarSaleSearchDescResponse {
        val carSale: CarSale = carSaleSearchUseCase.findCarSaleById(carId)
        httpServletResponse.addCookie(carSaleViewUseCase.increaseViewCount(carSale.carId, httpServletRequest))
        return carSaleSearchUseCase.searchCarSaleDescription(carSale)
    }

    fun searchDetailCars(
        pageable: Pageable,
        manu: String?,
        model: String?,
        submodel: String?,
        grade: String?
    ): Page<CarSearchDetailResponse> {
        return carSaleSearchUseCase.searchCarSaleDetail(pageable, manu, model, submodel, grade)
    }

    fun searchTagsCars(
        page:Int,
        size:Int,
        tag: String,
    ): List<CarSearchResponse> {
        val vector = embeddingPort.generateEmbedding(listOf(tag))
        val idList =  pineconePort.findTagsId(vector, page, size)
        val idOrderMap = idList.withIndex().associate { it.value to it.index }

        val unorderedResults = carSaleSearchUseCase.findCarSaleByTags(idList)
        return unorderedResults.sortedBy { idOrderMap[it.carId] }
    }
}